### PR TITLE
storage: resumption frontier is empty for closed collections

### PR DESCRIPTION
**Pull request by sploiselle**
tl;dr empty sinces need to stop dataflows from starting

### Motivation

This PR fixes a previously unreported bug.

We currently only downgrade the read handle to the empty antichain to signal that a collection has been dropped (see https://github.com/MaterializeInc/materialize/issues/17320#issuecomment-1441885572). However, if we are in a state where we drop a source and `clusterd` crashes and gets quickly rescheduled, we might try to create a source whose read handle is closed, though its write handle will still be < [].

This can cause a panic for the remap operator:
- [Its `as_of` is calculated to be `upper - 1`](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/source_reader_pipeline.rs#L708)
- [Its `since` can be empty](https://github.com/MaterializeInc/materialize/blob/main/src/storage-client/src/controller.rs#L1732-L1742), even when the upper is some value
- [On creation, it fails the assertion that its since is less than its as of when the since is empty](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/reclock/compat.rs#L114-L120)

To prevent this from occurring, I suggest we calculate the resumption frontier not only as the write handles upper, but as the join of the write and read handle, ensuring that if the read handle is closed, the resumption frontier will be empty, which will [stop the dataflow from being spawned on the rescheduled clusterd](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/storage_state.rs#L699).

This seems challenging to manufacture, but we have [observed it in production](https://materializeinc.sentry.io/issues/3993118310/events/a0815bc36b7147d29def78a2908a56b0/?project=6780145).

Note that implementing #17320 does not solve this condition absolutely because downgrading the write handle is not necessarily done before clusterd can be rescheduled.

I have a few open questions:
- Should this read handle be critical? It seems like the idea is that this is state one could hold onto, but it doesn't seem like it's used in that way.
- Is `join` the right semantics for this? Could also check if the read handle's since empty and continue; not sure what the aesthetic preference there is.

### Tips for reviewer

This is just a WIP for resolving this issue; this shouldn't be alarming and I'm putting something up only because it's Friday night and I didn't want to lose the fix's context.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by sploiselle** [comment]
I'm also working on a more detailed writeup of the incident, but that won't be done until Monday.

**Comment by benesch** [comment]
Wait, hang on! If the source is dropped... why is the controller telling the reincarnated cluster to reinstall it?

Is it possible the problem is that the rehydration logic is not correctly forgetting about dropped sources/sinks? Here's a quick diff to explain what I mean:

```diff
diff --git a/src/storage-client/src/controller/rehydration.rs b/src/storage-client/src/controller/rehydration.rs
index 796330e011..287f9d8151 100644
--- a/src/storage-client/src/controller/rehydration.rs
+++ b/src/storage-client/src/controller/rehydration.rs
@@ -415,12 +415,21 @@ where
             }
             StorageCommand::AllowCompaction(frontiers) => {
                 for (id, frontier) in frontiers {
-                    match self.sinks.get_mut(id) {
-                        Some(export) => {
-                            export.description.as_of.downgrade(frontier);
+                    if frontier.is_empty() {
+                        self.sinks
+                            .remove(id)
+                            .or_else(|| self.sources.remove(id))
+                            .unwrap_or_else(|| {
+                                panic!("AllowCompaction command for non-existent {id}")
+                            });
+                    } else {
+                        match self.sinks.get_mut(id) {
+                            Some(export) => {
+                                export.description.as_of.downgrade(frontier);
+                            }
+                            None if self.sources.contains_key(id) => continue,
+                            None => panic!("AllowCompaction command for non-existent {id}"),
                         }
-                        None if self.sources.contains_key(id) => continue,
-                        None => panic!("AllowCompaction command for non-existent {id}"),
                     }
                 }
             }
```

Totally untested. @aljoscha or @guswynn or whoever the owner of this code is these days should chime in. I might be on a wild goose chase.

**Comment by sploiselle** [comment]
Ah--this makes a lot of sense. I had not realized the code that brought storage processes back up to the current state was here. I had mistakenly believed it was brought back up to speed as a function of `storage_state::Worker::run_client`. Seems like your fix should definitely go in.

However, there is still a possible race condition here: dropping the collection, which downgrades the persist read handle's `since` is done at a different impedance than propagating the `AllowCompaction` command, i.e. one is done by a `PersistReadWorker` and the other is done by `storage_client::controller::Controller::process`. If you downgraded the collection's since and the worker restarted before `process` was called, we would still encounter the same panic. I'm not sure of anything that guarantees that ordering cannot occur.

**Review by philip-stoev — APPROVED** [review]
Given the convoluted description, I assume this is not easy to write a test for , right?

Please do run Nightly on this branch just in case.

**Comment by aljoscha** [comment]
> Wait, hang on! If the source is dropped... why is the controller telling the reincarnated cluster to reinstall it?
> 
> Is it possible the problem is that the rehydration logic is not correctly forgetting about dropped sources/sinks? Here's a quick diff to explain what I mean:

That is correct! It's a known problem and Petros and I have been discussing what the actual fix should look like. There is an [open PR](https://github.com/MaterializeInc/materialize/pull/17559) that _might_ have prevented this, but there are still concerns (from my side) about race conditions on the `clusterd` side. It's unfortunate that you had to debug this (and that I was out on this day), we should have resolved this earlier already! 🕊️ 

Roughly, the two positions on the topic are:
 - _state should change as soon as an intermediate component learns about it_: In this here case, it would mean dropping sources/sink from rehydration state as soon as we see the "outgoing" `AllowCompaction` message
 - _state should only change once we hear back from the component that we're forwarding to_: In the case here, it would mean (and does currently mean), that sources and sinks are dropped once we hear back a `DroppedIds` message from the `clusterd`, which it sends once it has seen the `AllowCompaction` message, but after jumping through some hoops first.

Nikhils suggestion would basically fix it "my way", why Petros' PR would fix it, by making sure we don't forget to keep sending the `AllowCompaction` message to the `clusterd`. I think we could embellish Petros' fix slightly by making sure that we patch up the `resume_upper` of `CreateSourceCommands` once we see compaction to the empty frontier. This together with parts of the PR here would make sure we don't try and render these sources. Basically, we would consolidate the `CreateSourceCommands` and `AllowCompaction` messages.
